### PR TITLE
Create docs transport package and turn on extensions source of truth

### PR DESCRIPTION
**Pull request by ViktorHofer**
Contributes to https://github.com/dotnet/runtime/issues/996

- Create docs ref+xml transport package
Add a nuget package that contains all the reference assemblies and
source-of-truth API docs XML files for the current release.

- Turn on source of truth for Microsoft.Extensions.*
Make the source of truth for API docs for the Microsoft.Extensions.*
libraries dotnet/runtime instead of the intellisense package.
Disable the few projects that aren't yet documented. Same for libraries
that are already effectively source-of-truth in runtime but which aren't
documented.

@carlossanlop there are ~20 remaining undocumented APIs in all Microsoft.Extensions.*. Can you please help resolve them as part of this PR?

```
C:\git\runtime2\src\libraries\Microsoft.Extensions.Primitives\src\InplaceStringBuilder.cs(14,19): error CS1591: Missing XML comment for publicly visible type or member 'InplaceStringBuilder' [C:\git\runtime2\src\libraries\Microsoft.Extensions.Primitives\src\Microsoft.Extensions.Primitives.csproj::TargetFramework=net8.0]
C:\git\runtime2\src\libraries\Microsoft.Extensions.Primitives\src\InplaceStringBuilder.cs(20,16): error CS1591: Missing XML comment for publicly visible type or member 'InplaceStringBuilder.InplaceStringBuilder(int)' [C:\git\runtime2\src\libraries\Microsoft.Extensions.Primitives\src\Microsoft.Extensions.Primitives.csproj::TargetFramework=net8.0]
C:\git\runtime2\src\libraries\Microsoft.Extensions.Primitives\src\InplaceStringBuilder.cs(30,20): error CS1591: Missing XML comment for publicly visible type or member 'InplaceStringBuilder.Capacity' [C:\git\runtime2\src\libraries\Microsoft.Extensions.Primitives\src\Microsoft.Extensions.Primitives.csproj::TargetFramework=net8.0]
C:\git\runtime2\src\libraries\Microsoft.Extensions.Primitives\src\InplaceStringBuilder.cs(50,21): error CS1591: Missing XML comment for publicly visible type or member 'InplaceStringBuilder.Append(string?)' [C:\git\runtime2\src\libraries\Microsoft.Extensions.Primitives\src\Microsoft.Extensions.Primitives.csproj::TargetFramework=net8.0]
C:\git\runtime2\src\libraries\Microsoft.Extensions.Primitives\src\InplaceStringBuilder.cs(60,21): error CS1591: Missing XML comment for publicly visible type or member 'InplaceStringBuilder.Append(StringSegment)' [C:\git\runtime2\src\libraries\Microsoft.Extensions.Primitives\src\Microsoft.Extensions.Primitives.csproj::TargetFramework=net8.0]
C:\git\runtime2\src\libraries\Microsoft.Extensions.Primitives\src\InplaceStringBuilder.cs(66,28): error CS1591: Missing XML comment for publicly visible type or member 'InplaceStringBuilder.Append(string?, int, int)' [C:\git\runtime2\src\libraries\Microsoft.Extensions.Primitives\src\Microsoft.Extensions.Primitives.csproj::TargetFramework=net8.0]
C:\git\runtime2\src\libraries\Microsoft.Extensions.Primitives\src\InplaceStringBuilder.cs(87,28): error CS1591: Missing XML comment for publicly visible type or member 'InplaceStringBuilder.Append(char)' [C:\git\runtime2\src\libraries\Microsoft.Extensions.Primitives\src\Microsoft.Extensions.Primitives.csproj::TargetFramework=net8.0]
C:\git\runtime2\src\libraries\Microsoft.Extensions.Primitives\src\InplaceStringBuilder.cs(102,33): error CS1591: Missing XML comment for publicly visible type or member 'InplaceStringBuilder.ToString()' [C:\git\runtime2\src\libraries\Microsoft.Extensions.Primitives\src\Microsoft.Extensions.Primitives.csproj::TargetFramework=net8.0]
C:\git\runtime2\src\libraries\Microsoft.Extensions.Primitives\src\StringSegmentComparer.cs(35,20): error CS1591: Missing XML comment for publicly visible type or member 'StringSegmentComparer.Compare(StringSegment, StringSegment)' [C:\git\runtime2\src\libraries\Microsoft.Extensions.Primitives\src\Microsoft.Extensions.Primitives.csproj::TargetFramework=net8.0]
C:\git\runtime2\src\libraries\Microsoft.Extensions.Primitives\src\StringSegmentComparer.cs(40,21): error CS1591: Missing XML comment for publicly visible type or member 'StringSegmentComparer.Equals(StringSegment, StringSegment)' [C:\git\runtime2\src\libraries\Microsoft.Extensions.Primitives\src\Microsoft.Extensions.Primitives.csproj::TargetFramework=net8.0]
C:\git\runtime2\src\libraries\Microsoft.Extensions.Primitives\src\StringTokenizer.cs(99,34): error CS1591: Missing XML comment for publicly visible type or member 'StringTokenizer.Enumerator.Current' [C:\git\runtime2\src\libraries\Microsoft.Extensions.Primitives\src\Microsoft.Extensions.Primitives.csproj::TargetFramework=net8.0]
C:\git\runtime2\src\libraries\Microsoft.Extensions.Primitives\src\StringTokenizer.cs(103,25): error CS1591: Missing XML comment for publicly visible type or member 'StringTokenizer.Enumerator.Dispose()' [C:\git\runtime2\src\libraries\Microsoft.Extensions.Primitives\src\Microsoft.Extensions.Primitives.csproj::TargetFramework=net8.0]
C:\git\runtime2\src\libraries\Microsoft.Extensions.Primitives\src\StringTokenizer.cs(107,25): error CS1591: Missing XML comment for publicly visible type or member 'StringTokenizer.Enumerator.MoveNext()' [C:\git\runtime2\src\libraries\Microsoft.Extensions.Primitives\src\Microsoft.Extensions.Primitives.csproj::TargetFramework=net8.0]
C:\git\runtime2\src\libraries\Microsoft.Extensions.Primitives\src\StringTokenizer.cs(128,25): error CS1591: Missing XML comment for publicly visible type or member 'StringTokenizer.Enumerator.Reset()' [C:\git\runtime2\src\libraries\Microsoft.Extensions.Primitives\src\Microsoft.Extensions.Primitives.csproj::TargetFramework=net8.0]
C:\git\runtime2\src\libraries\Microsoft.Extensions.Primitives\src\StringValues.cs(791,25): error CS1591: Missing XML comment for publicly visible type or member 'StringValues.Enumerator.MoveNext()' [C:\git\runtime2\src\libraries\Microsoft.Extensions.Primitives\src\Microsoft.Extensions.Primitives.csproj::TargetFramework=net8.0]
C:\git\runtime2\src\libraries\Microsoft.Extensions.Primitives\src\StringValues.cs(817,28): error CS1591: Missing XML comment for publicly visible type or member 'StringValues.Enumerator.Current' [C:\git\runtime2\src\libraries\Microsoft.Extensions.Primitives\src\Microsoft.Extensions.Primitives.csproj::TargetFramework=net8.0]
C:\git\runtime2\src\libraries\Microsoft.Extensions.Primitives\src\StringValues.cs(826,25): error CS1591: Missing XML comment for publicly visible type or member 'StringValues.Enumerator.Dispose()' [C:\git\runtime2\src\libraries\Microsoft.Extensions.Primitives\src\Microsoft.Extensions.Primitives.csproj::TargetFramework=net8.0]

C:\git\runtime2\src\libraries\Microsoft.Extensions.DependencyInjection.Abstractions\src\FromKeyedServicesAttribute.cs(9,18): error CS1591: Missing XML comment for publicly visible type or member 'FromKeyedServicesAttribute' [C:\git\runtime2\src\libraries\Microsoft.Extensions.DependencyInjection.Abstractions\src\Microsoft.Extensions.DependencyInjection.Abstractions.csproj::TargetFramework=net8.0]
C:\git\runtime2\src\libraries\Microsoft.Extensions.DependencyInjection.Abstractions\src\FromKeyedServicesAttribute.cs(11,16): error CS1591: Missing XML comment for publicly visible type or member 'FromKeyedServicesAttribute.FromKeyedServicesAttribute(object)' [C:\git\runtime2\src\libraries\Microsoft.Extensions.DependencyInjection.Abstractions\src\Microsoft.Extensions.DependencyInjection.Abstractions.csproj::TargetFramework=net8.0]
C:\git\runtime2\src\libraries\Microsoft.Extensions.DependencyInjection.Abstractions\src\FromKeyedServicesAttribute.cs(13,23): error CS1591: Missing XML comment for publicly visible type or member 'FromKeyedServicesAttribute.Key' [C:\git\runtime2\src\libraries\Microsoft.Extensions.DependencyInjection.Abstractions\src\Microsoft.Extensions.DependencyInjection.Abstractions.csproj::TargetFramework=net8.0]
C:\git\runtime2\src\libraries\Microsoft.Extensions.DependencyInjection.Abstractions\src\IKeyedServiceProvider.cs(8,22): error CS1591: Missing XML comment for publicly visible type or member 'IKeyedServiceProvider' [C:\git\runtime2\src\libraries\Microsoft.Extensions.DependencyInjection.Abstractions\src\Microsoft.Extensions.DependencyInjection.Abstractions.csproj::TargetFramework=net8.0]
C:\git\runtime2\src\libraries\Microsoft.Extensions.DependencyInjection.Abstractions\src\IKeyedServiceProvider.cs(29,25): error CS1591: Missing XML comment for publicly visible type or member 'KeyedService' [C:\git\runtime2\src\libraries\Microsoft.Extensions.DependencyInjection.Abstractions\src\Microsoft.Extensions.DependencyInjection.Abstractions.csproj::TargetFramework=net8.0]
C:\git\runtime2\src\libraries\Microsoft.Extensions.DependencyInjection.Abstractions\src\IKeyedServiceProvider.cs(31,30): error CS1591: Missing XML comment for publicly visible type or member 'KeyedService.AnyKey' [C:\git\runtime2\src\libraries\Microsoft.Extensions.DependencyInjection.Abstractions\src\Microsoft.Extensions.DependencyInjection.Abstractions.csproj::TargetFramework=net8.0]
C:\git\runtime2\src\libraries\Microsoft.Extensions.DependencyInjection.Abstractions\src\IServiceProviderIsKeyedService.cs(8,22): error CS1591: Missing XML comment for publicly visible type or member 'IServiceProviderIsKeyedService' [C:\git\runtime2\src\libraries\Microsoft.Extensions.DependencyInjection.Abstractions\src\Microsoft.Extensions.DependencyInjection.Abstractions.csproj::TargetFramework=net8.0]
C:\git\runtime2\src\libraries\Microsoft.Extensions.DependencyInjection.Abstractions\src\ServiceDescriptor.cs(250,21): error CS1591: Missing XML comment for publicly visible type or member 'ServiceDescriptor.IsKeyedService' [C:\git\runtime2\src\libraries\Microsoft.Extensions.DependencyInjection.Abstractions\src\Microsoft.Extensions.DependencyInjection.Abstractions.csproj::TargetFramework=net8.0]
C:\git\runtime2\src\libraries\Microsoft.Extensions.DependencyInjection.Abstractions\src\ServiceKeyAttribute.cs(9,18): error CS1591: Missing XML comment for publicly visible type or member 'ServiceKeyAttribute' [C:\git\runtime2\src\libraries\Microsoft.Extensions.DependencyInjection.Abstractions\src\Microsoft.Extensions.DependencyInjection.Abstractions.csproj::TargetFramework=net8.0]
```

### CODE REVIEW

**Comment by ViktorHofer** [comment]
@dotnet/area-extensions-dependencyinjection @dotnet/area-extensions-primitives can you please help annotating the above listed APIs as part of this PR?

File: eng/intellisense.targets [hunk 1]
```diff
@@ -1,40 +1,46 @@
-<Project InitialTargets="VerifyAssemblySupportsDocsXmlGeneration">
+<Project>
 
   <PropertyGroup>
-    <UseIntellisensePackageDocXmlFile Condition="'$(UseIntellisensePackageDocXmlFile)' == ''">true</UseIntellisensePackageDocXmlFile>
+    <UseCompilerGeneratedDocXmlFile Condition="'$(UseCompilerGeneratedDocXmlFile)' == '' and $(MSBuildProjectName.StartsWith('Microsoft.Extensions.'))">true</UseCompilerGeneratedDocXmlFile>
   </PropertyGroup>
 
-  <Target Name="VerifyAssemblySupportsDocsXmlGeneration"
-          Condition="'$(UseIntellisensePackageDocXmlFile)' != 'true'">
```
Comment (carlossanlop) [inline eng/intellisense.targets:8]: Let me see if I understand this: You're now inverting the the check for `UseCompilerGeneratedDocXmlFile` in line 7 (before line 15) which means we can disable simply skip the compiler-generated doc xml file logic in parital facade assemblies and in PNSE assemblies. Correct?

File: src/libraries/oob-src.proj [hunk 2]
```diff
@@ -32,7 +32,8 @@
 
     <!-- Don't build meta-projects in the NetCoreAppCurrent vertical. -->
     <ProjectReference Remove="Microsoft.Internal.Runtime.AspNetCore.Transport\src\Microsoft.Internal.Runtime.AspNetCore.Transport.proj;
-                              Microsoft.Internal.Runtime.WindowsDesktop.Transport\src\Microsoft.Internal.Runtime.WindowsDesktop.Transport.proj" />
+                              Microsoft.Internal.Runtime.WindowsDesktop.Transport\src\Microsoft.Internal.Runtime.WindowsDesktop.Transport.proj;
+                              Microsoft.Internal.Runtime.DotNetApiDocs.Transport\src\Microsoft.Internal.Runtime.DotNetApiDocs.Transport.proj" />
```
Comment (carlossanlop) [inline src/libraries/oob-src.proj:36]: Ok I see: we only build these in all configurations, but not in src.

Questions:

- The existing transport project for aspnetcore already includes ref assemblies and xmls?
- I don't see the windowsdesktop one here, but I assume it already existed. Does also include xmls?

File: src/libraries/Microsoft.Internal.Runtime.DotNetApiDocs.Transport/src/Microsoft.Internal.Runtime.DotNetApiDocs.Transport.proj [hunk 3]
```diff
@@ -0,0 +1,52 @@
+<Project Sdk="Microsoft.Build.NoTargets">
+
+  <PropertyGroup>
+    <TargetFramework>$(NetCoreAppCurrent)-$(TargetOS)</TargetFramework>
+    <IsShipping>false</IsShipping>
+    <!-- Reference the outputs to have them available as build outputsd. -->
```
Comment (carlossanlop) [inline src/libraries/Microsoft.Internal.Runtime.DotNetApiDocs.Transport/src/Microsoft.Internal.Runtime.DotNetApiDocs.Transport.proj:6]:
```suggestion
    <!-- Reference the outputs to have them available as build outputs. -->
```

File: src/libraries/Microsoft.Internal.Runtime.DotNetApiDocs.Transport/src/Microsoft.Internal.Runtime.DotNetApiDocs.Transport.proj [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    <NoTargetsDoNotReferenceOutputAssemblies>false</NoTargetsDoNotReferenceOutputAssemblies>
+    <IsPackable>true</IsPackable>
```
Comment (carlossanlop) [inline src/libraries/Microsoft.Internal.Runtime.DotNetApiDocs.Transport/src/Microsoft.Internal.Runtime.DotNetApiDocs.Transport.proj:8]: When the `release/8.0` branch becomes servicing, I assume we will disable this transport package, right?

File: src/libraries/Microsoft.Internal.Runtime.DotNetApiDocs.Transport/src/Microsoft.Internal.Runtime.DotNetApiDocs.Transport.proj [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+                                       @(CompilerGeneratedXmlDocFile)"
+                              PackagePath="ref\$(NetCoreAppCurrent)\" />
+    </ItemGroup>
+
+    <ItemGroup>
+      <PartialFacadeCompilerGeneratedXmlDocFile Include="@(CompilerGeneratedXmlDocFile->WithMetadataValue('IsPartialFacadeAssembly', 'true'))" />
+      <PlatformNotSupportedCompilerGeneratedXmlDocFile Include="@(CompilerGeneratedXmlDocFile->WithMetadataValue('IsPlatformNotSupportedAssembly', 'true'))" />
+    </ItemGroup>
+
+    <Error Text="Compiler generated XML documentation files are missing data because partial facade assemblies aren't yet supported by the repo infrastructure. Consider removing the partial facade feature or setting the 'GenerateDocumentationFile' property to false in the project file. Assemblies: @(PartialFacadeCompilerGeneratedXmlDocFile->Metadata('Filename'), ', ')."
```
Comment (carlossanlop) [inline src/libraries/Microsoft.Internal.Runtime.DotNetApiDocs.Transport/src/Microsoft.Internal.Runtime.DotNetApiDocs.Transport.proj:45]: Nice, you moved the errors here.

**Comment by tarekgh** [comment]
@benjaminpetit @adityamandaleeka could one of you adding the missing triple slash docs to the keyed service APIs you have added? The list of the missing docs is mentioned in the issue description above. Thanks!

**Comment by adityamandaleeka** [comment]
@tarekgh @benjaminpetit I just pushed the Keyed DI ones onto this PR.

**Comment by carlossanlop** [comment]
The remaining APIs are ConsoleLoggerSettings from Microsoft.Extensions.ConsoleLoggerSettings. They are marked as obsolete: https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.logging.console.consoleloggersettings?view=dotnet-plat-ext-2.2

I can add the documentation for those.

**Comment by ViktorHofer** [comment]
Hmm, now there are a few undocumented `Microsoft.Extensions.Hosting` APIs showing up. I think that it initially didn't show up as the compiler didn't attempt to build it because of a failed dependency in the build graph.

**Comment by carlossanlop** [comment]
> Hmm, now there are a few undocumented `Microsoft.Extensions.Hosting` APIs showing up. I think that it initially didn't show up as the compiler didn't attempt to build it because of a failed dependency in the build graph.

Let me try to document them. Otherwise, I'll ask the author to help.

**Comment by carlossanlop** [comment]
> Let me try to document them. Otherwise, I'll ask the author to help.

Done. I also built `libs` locally and I didn't get any more build failures. Crossing fingers.

**Comment by carlossanlop** [comment]
New error:
```
src/libraries/Microsoft.Internal.Runtime.DotNetApiDocs.Transport/src/
Microsoft.Internal.Runtime.DotNetApiDocs.Transport.proj(45,5):
error : (NETCORE_ENGINEERING_TELEMETRY=Build)
Compiler generated XML documentation files are missing data because partial facade assemblies aren't yet supported 
by the repo infrastructure. Consider removing the partial facade feature or setting the 'GenerateDocumentationFile' 
property to false in the project file. 
Assemblies: Microsoft.Bcl.AsyncInterfaces, Microsoft.Bcl.Cryptography, Microsoft.Bcl.TimeProvider.

```

File: src/libraries/oob-src.proj (hunk 2 above)
Comment (ViktorHofer) [reply src/libraries/oob-src.proj:36]: > I don't see the windowsdesktop one here, but I assume it already existed.

It's in the list above. It includes ref assemblies, lib assemblies and xml docs.

> The existing transport project for aspnetcore already includes ref assemblies and xmls?

Yes, the aspnetcore package includes ref assemblies, lib assemblies and xml docs.

File: src/libraries/Microsoft.Internal.Runtime.DotNetApiDocs.Transport/src/Microsoft.Internal.Runtime.DotNetApiDocs.Transport.proj (hunk 4 above)
Comment (ViktorHofer) [reply src/libraries/Microsoft.Internal.Runtime.DotNetApiDocs.Transport/src/Microsoft.Internal.Runtime.DotNetApiDocs.Transport.proj:8]: dotnet/runtime uses a incremental servicing model which means that this package won't be generated anymore after the RTM release.

**Review by carlossanlop — APPROVED** [review]
:shipit:  🚢 🚀

File: src/libraries/Microsoft.Extensions.Caching.Abstractions/src/CacheEntryExtensions.cs [hunk 6]
```diff
@@ -7,6 +7,9 @@
 
 namespace Microsoft.Extensions.Caching.Memory
 {
+    /// <summary>
+    /// Provide extensions methods for <see cref="ICacheEntry"/> operations.
```
Comment (xakep139) [inline src/libraries/Microsoft.Extensions.Caching.Abstractions/src/CacheEntryExtensions.cs:11]: `extension methods` maybe?

**Pull request merged**